File: dsa/6.hash_table/560_Subarray_Sum_Equals_K.java
package hash_table;
import java.util.*;

class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> mapVal = new HashMap<Integer, Integer>();
        int cumSum = 0;
        int result = 0;

        for (int i = 0; i < nums.length; i++){
            cumSum += nums[i];
            if (cumSum == k){
                System.out.println(cumSum);
                result += 1;
            }

            if (mapVal.containsKey(cumSum - k) == true){
                result += mapVal.get(cumSum - k);
            }

            if(mapVal.containsKey(cumSum) == true){
                mapVal.replace(cumSum, mapVal.get(cumSum) + 1);
            }
            else{
                mapVal.put(cumSum, 1);
            }


        }
        return result;
    }
}

File: dsa/6.hash_table/560_Subarray_Sum_Equals_K.py
import typing as t

class Solution:
    def subarraySum(self, nums: t.List[int], k: int) -> int:
        result: int = 0
        acc_sum: int = 0
        prefix_sum_dict: t.Dict[int, int] = {
            0: 1
        }

        for ind in range(0, len(nums), 1):
            acc_sum += nums[ind]
            if (acc_sum - k) in prefix_sum_dict.keys():
                result += prefix_sum_dict.get(acc_sum - k)

            prefix_sum_dict[acc_sum] = prefix_sum_dict.get(acc_sum, 0) + 1
        
        return result